File: src/main/java/com/joraph/Query.java
package com.joraph;

import static java.util.Arrays.asList;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class Query {

	private Set<Class<?>> entityClasses;
	private Set<?> rootObjects;
	private ObjectGraph existingGraph;
	private List<Object> arguments = new ArrayList<>();

	public Query() {
		
	}

	public Query(Class<?>... entityClasses) {
		withEntityClasses(Arrays.stream(entityClasses)
				.collect(Collectors.toSet()));
	}


	/**
	 * @return the entityClasses
	 */
	public Set<Class<?>> getEntityClasses() {
		return entityClasses;
	}

	/**
	 * @param entityClasses the entityClasses to set
	 */
	public Query withEntityClasses(Set<Class<?>> entityClasses) {
		this.entityClasses = entityClasses;
		return this;
	}

	/**
	 * @param entityClass the entityClass to set
	 */
	public Query withEntityClass(Class<?> entityClass) {
		if (this.entityClasses==null || this.entityClasses.isEmpty()) {
			this.entityClasses = new HashSet<Class<?>>();
		}
		this.entityClasses.add(entityClass);
		return this;
	}

	/**
	 * @return the argumentProvider
	 */
	public List<Object> getArguments() {
		return arguments;
	}

	/**
	 * @return the rootObjects
	 */
	public Set<?> getRootObjects() {
		return rootObjects;
	}

	/**
	 * @param rootObjects the rootObjects to set
	 */
	public Query withRootObject(Object... rootObjects) {
		return withRootObjects(Arrays.stream(rootObjects)
				.collect(Collectors.toSet()));
	}

	/**
	 * @param rootObjects the rootObjects to set
	 */
	public Query withRootObjects(Iterable<?> rootObjects) {
		this.rootObjects = CollectionUtil.toSet(rootObjects);
		return this;
	}

	/**
	 * @return the existingGraph
	 */
	public ObjectGraph getExistingGraph() {
		return existingGraph;
	}

	/**
	 * @return whether or not there is an existing graph
	 */
	public boolean hasExistingGraph() {
		return existingGraph!=null;
	}

	/**
	 * @param existingGraph the existingGraph to set
	 */
	public Query withExistingGraph(ObjectGraph existingGraph) {
		this.existingGraph = existingGraph;
		return this;
	}

	/**
	 * @param arguments
	 */
	public Query withArguments(List<Object> arguments) {
		this.arguments = new ArrayList<>(arguments);
		return this;
	}

	/**
	 * @param arguments
	 */
	public Query withArgument(Object argument) {
		this.arguments = new ArrayList<>();
		this.arguments.add(argument);
		return this;
	}

	/**
	 * @param arguments
	 */
	public Query addArguments(List<Object> arguments) {
		this.arguments.addAll(arguments);
		return this;
	}

	/**
	 * @param arguments
	 */
	public Query addArgument(Object argument) {
		this.arguments.add(argument);
		return this;
	}

	/**
	 * @param arguments
	 */
	public Query addArguments(Object... arguments) {
		this.arguments.addAll(asList(arguments));
		return this;
	}

}

File: src/main/kotlin/com/joraph/Query.kt
package com.joraph

/**
 * A [Query] is used to define which objects in a graph to load. A [Query] starts
 * with one or more [rootObjects] from which their [com.joraph.schema.Schema] is consulted to determine
 * what additional objects in their graph should be loaded, the same logic is applied
 * to every object that is loaded by the [Query] until the entire graph has been
 * loaded.
 */
open class Query {

    val entityClasses: MutableSet<Class<*>> = mutableSetOf()
    val rootObjects: MutableSet<Any> = mutableSetOf()
    val arguments: MutableList<Any> = mutableListOf()

    var existingGraph: ObjectGraph? = null
        private set

    val hasExistingGraph get() = existingGraph != null

    constructor()
    constructor(vararg entityClasses: Class<*>) {
        withEntityClass(*entityClasses)
    }

    /**
     * Instructs the [Query] to load the dependencies of all of the given [entityClasses]
     * present in the [ObjectGraph].
     */
    fun withEntityClasses(entityClasses: Collection<Class<*>>): Query {
        this.entityClasses.addAll(entityClasses)
        return this
    }

    /**
     * Instructs the [Query] to load the dependencies of all of the given [entityClasses]
     * present in the [ObjectGraph].
     */
    fun withEntityClass(vararg entityClasses: Class<*>): Query {
        return withEntityClasses(entityClasses.toList())
    }

    /**
     * Instructs the [Query] to load the dependencies of all of the given [rootEntities]
     * and to add them and their dependencies to the [ObjectGraph].
     */
    fun withRootEntities(rootEntities: Collection<Any>): Query {
        return withRootObjects(rootEntities)
            .withEntityClasses(rootEntities.map { it::class.java })
    }

    /**
     * Instructs the [Query] to load the dependencies of all of the given [rootEntities]
     * and to add them and their dependencies to the [ObjectGraph].
     */
    fun withRootEntity(vararg rootEntities: Any): Query {
        return withRootObjects(rootEntities.toList())
            .withEntityClasses(rootEntities.map { it::class.java })
    }

    /**
     * Instructs the [Query] to add the given [rootObjects] to the [ObjectGraph].
     * This does not cause their dependencies to be loaded, for that use
     * [withRootEntities].
     */
    fun withRootObjects(rootObjects: Collection<Any>): Query {
        this.rootObjects.addAll(rootObjects)
        return this
    }

    /**
     * Instructs the [Query] to add the given [rootObjects] to the [ObjectGraph].
     * This does not cause their dependencies to be loaded, for that use
     * [withRootEntity].
     */
    fun withRootObject(vararg rootObjects: Any): Query {
        this.rootObjects.addAll(rootObjects.toList())
        return this
    }

    /**
     * Instructs the [Query] to put results into the given [ObjectGraph].
     */
    fun withExistingGraph(existingGraph: ObjectGraph?): Query {
        this.existingGraph = existingGraph
        return this
    }

    /**
     * Adds objects that can provide additional arguments to [com.joraph.loader.LoaderFunction]s
     * when executing the [Query].
     */
    fun withArgumentProviders(arguments: Collection<Any>): Query {
        this.arguments.addAll(arguments)
        return this
    }

    /**
     * Adds objects that can provide additional arguments to [com.joraph.loader.LoaderFunction]s
     * when executing the [Query].
     */
    fun withArgumentProvider(vararg arguments: Any): Query {
        return withArgumentProviders(arguments.toList())
    }

    @Deprecated(
        message = "Use withArgumentProviders",
        replaceWith = ReplaceWith("withArgumentProviders"))
    fun withArguments(arguments: Collection<Any>): Query {
        return withArgumentProviders(arguments)
    }

    @Deprecated(
        message = "Use withArgumentProvider",
        replaceWith = ReplaceWith("withArgumentProvider"))
    fun withArgument(vararg arguments: Any): Query {
        return withArgumentProviders(arguments.toList())
    }
}
